cdc: panic after disable syncpoint through cli update

### What did you do?

create changefeed with syncpoint enable
pause 
update with syncpoint disable
resume

### What did you expect to see?

not panic

### What did you see instead?

cdc panic instead
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x2eac771]

goroutine 543 [running]:
github.com/pingcap/tiflow/cdc/owner.(*ddlSinkImpl).emitSyncPoint(0xc0009ae690, 0x41031c8, 0xc000b75650, 0x5fe5c9046dc0000, 0x3093e9c0, 0xed9c4f221)
	github.com/pingcap/tiflow/cdc/owner/ddl_sink.go:205 +0x71
github.com/pingcap/tiflow/cdc/owner.(*changefeed).handleBarrier(0xc002448a20, 0x41031c8, 0xc000b75650, 0x5fe5c9046dc0000, 0x0, 0x0)
	github.com/pingcap/tiflow/cdc/owner/changefeed.go:464 +0x1bc
github.com/pingcap/tiflow/cdc/owner.(*changefeed).tick(0xc002448a20, 0x41031c8, 0xc000b75650, 0xc00249daa0, 0xc001b22b70, 0x2eb20ba, 0xc00335cf30)
	github.com/pingcap/tiflow/cdc/owner/changefeed.go:178 +0x24d
github.com/pingcap/tiflow/cdc/owner.(*changefeed).Tick(0xc002448a20, 0x41030d8, 0xc000b75638, 0xc00249daa0, 0xc001b22b70)
	github.com/pingcap/tiflow/cdc/owner/changefeed.go:117 +0x149
github.com/pingcap/tiflow/cdc/owner.(*Owner).Tick(0xc000613030, 0x7f19e2a75cf0, 0xc000ba08a0, 0x40af638, 0xc000916440, 0x1, 0x0, 0x1, 0x0)
	github.com/pingcap/tiflow/cdc/owner/owner.go:177 +0x36d
github.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run(0xc0004b3780, 0x7f19e2a75cf0, 0xc000ba08a0, 0xc001b22900, 0xbebc200, 0x7fff617b0e66, 0x12, 0x0, 0x0)
	github.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:239 +0x591
github.com/pingcap/tiflow/cdc/capture.(*Capture).runEtcdWorker(0xc001e20000, 0x4103178, 0xc000ba08a0, 0x40728a0, 0xc000613030, 0x40af638, 0xc000916440, 0xbebc200, 0x0, 0x0)
	github.com/pingcap/tiflow/cdc/capture/capture.go:450 +0x167
github.com/pingcap/tiflow/cdc/capture.(*Capture).campaignOwner(0xc001e20000, 0x4103178, 0xc00063ce40, 0x0, 0x0)
	github.com/pingcap/tiflow/cdc/capture/capture.go:428 +0xa25
github.com/pingcap/tiflow/cdc/capture.(*Capture).run.func2(0xc000f92100, 0xc001e20000, 0x4103178, 0xc00063ce40, 0xc001b800e0)
	github.com/pingcap/tiflow/cdc/capture/capture.go:292 +0xb5
created by github.com/pingcap/tiflow/cdc/capture.(*Capture).run
	github.com/pingcap/tiflow/cdc/capture/capture.go:286 +0x37b
```

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
5.4
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
5.4
```

--- a/cdc/owner/changefeed.go
+++ b/cdc/owner/changefeed.go
@@ -225,7 +225,11 @@ func (c *changefeed) tick(ctx cdcContext.Context, state *orchestrator.Changefeed
 		return errors.Trace(err)
 	default:
 	}
-
+	// we need to wait sink to be ready before we do the other things
+	// otherwise, we may cause a nil pointer panic
+	if !c.sink.isInitialized() {
+		return nil
+	}
 	// This means that the cached DDL has been executed,
 	// and we need to use the latest table names.
 	if c.currentTableNames == nil {

--- a/cdc/owner/changefeed_test.go
+++ b/cdc/owner/changefeed_test.go
@@ -139,6 +139,10 @@ func (m *mockDDLSink) close(ctx context.Context) error {
 	return nil
 }
 
+func (m *mockDDLSink) isInitialized() bool {
+	return true
+}
+
 func (m *mockDDLSink) Barrier(ctx context.Context) error {
 	return nil
 }

--- a/cdc/owner/ddl_sink.go
+++ b/cdc/owner/ddl_sink.go
@@ -16,6 +16,7 @@ package owner
 import (
 	"context"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/pingcap/errors"
@@ -53,6 +54,7 @@ type DDLSink interface {
 	emitSyncPoint(ctx cdcContext.Context, checkpointTs uint64) error
 	// close the sink, cancel running goroutine.
 	close(ctx context.Context) error
+	isInitialized() bool
 }
 
 type ddlSinkImpl struct {
@@ -79,16 +81,22 @@ type ddlSinkImpl struct {
 	// cancel would be used to cancel the goroutine start by `run`
 	cancel context.CancelFunc
 	wg     sync.WaitGroup
+	// we use `initialized` to indicate whether the sink has been initialized.
+	// the caller before calling any method of ddl sink
+	// should check `initialized` first
+	initialized atomic.Value
 }
 
 func newDDLSink() DDLSink {
-	return &ddlSinkImpl{
+	res := &ddlSinkImpl{
 		ddlSentTsMap:    make(map[*model.DDLEvent]uint64),
 		ddlCh:           make(chan *model.DDLEvent, 1),
 		errCh:           make(chan error, defaultErrChSize),
 		sinkInitHandler: ddlSinkInitializer,
 		cancel:          func() {},
 	}
+	res.initialized.Store(false)
+	return res
 }
 
 type ddlSinkInitHandler func(ctx cdcContext.Context, a *ddlSinkImpl, id model.ChangeFeedID, info *model.ChangeFeedInfo) error
@@ -114,6 +122,9 @@ func ddlSinkInitializer(ctx cdcContext.Context, a *ddlSinkImpl, id model.ChangeF
 	if err != nil {
 		return errors.Trace(err)
 	}
+	failpoint.Inject("DDLSinkInitializeSlowly", func() {
+		time.Sleep(time.Second * 5)
+	})
 	a.syncPointStore = syncPointStore
 
 	if err := a.syncPointStore.CreateSynctable(stdCtx); err != nil {
@@ -139,6 +150,7 @@ func (s *ddlSinkImpl) run(ctx cdcContext.Context, id model.ChangeFeedID, info *m
 			ctx.Throw(err)
 			return
 		}
+		s.initialized.Store(true)
 		log.Info("ddl sink initialized, start processing...",
 			zap.String("namespace", ctx.ChangefeedVars().ID.Namespace),
 			zap.String("changefeed", ctx.ChangefeedVars().ID.ID),
@@ -304,4 +316,8 @@ func (s *ddlSinkImpl) close(ctx context.Context) (err error) {
 		return err
 	}
 	return nil
+}
+
+func (s *ddlSinkImpl) isInitialized() bool {
+	return s.initialized.Load().(bool)
 }

--- a/tests/integration_tests/syncpoint/run.sh
+++ b/tests/integration_tests/syncpoint/run.sh
@@ -155,6 +155,8 @@ function run() {
 	run_sql "CREATE DATABASE testSync;"
 	run_sql "CREATE DATABASE testSync;" ${DOWN_TIDB_HOST} ${DOWN_TIDB_PORT}
 
+	# make suer no panic happen when the syncpoint enable and the ddl sink initializing slowly
+	export GO_FAILPOINTS='github.com/pingcap/tiflow/cdc/owner/DDLSinkInitializeSlowly=return(true)'
 	start_ts=$(run_cdc_cli_tso_query ${UP_PD_HOST_1} ${UP_PD_PORT_1})
 	run_cdc_server --workdir $WORK_DIR --binary $CDC_BINARY
 

